File: src/commonMain/kotlin/com/scoreo/application/GetPlayerStatsUseCase.kt
package com.scoreo.application

import com.scoreo.domain.port.GameTypeRepository
import com.scoreo.domain.port.MatchRepository

data class PlayerStats(
    val playerId: String,
    val wins: Int,
    val losses: Int,
)

class GetPlayerStatsUseCase(
    private val matchRepository: MatchRepository,
    private val gameTypeRepository: GameTypeRepository,
) {
    operator fun invoke(): Map<String, PlayerStats> {
        val stats = mutableMapOf<String, PlayerStats>()

        matchRepository.getAll().forEach { match ->
            val gameType = gameTypeRepository.findById(match.gameTypeId) ?: return@forEach
            val winners = match.getWinners(gameType).toSet()

            match.playerScores.forEach { ps ->
                val current = stats[ps.playerId] ?: PlayerStats(ps.playerId, 0, 0)
                stats[ps.playerId] = if (ps.playerId in winners) {
                    current.copy(wins = current.wins + 1)
                } else {
                    current.copy(losses = current.losses + 1)
                }
            }
        }
        return stats
    }
}

File: src/application/getPlayerStatsUseCase.ts
import { getWinners } from '../domain/model/match'
import type { GameTypeRepository } from '../domain/port/gameTypeRepository'
import type { MatchRepository } from '../domain/port/matchRepository'

export interface PlayerStats {
  playerId: string
  wins: number
  losses: number
}

export class GetPlayerStatsUseCase {
  constructor(
    private readonly matchRepository: MatchRepository,
    private readonly gameTypeRepository: GameTypeRepository,
  ) {}

  invoke(): Map<string, PlayerStats> {
    const stats = new Map<string, PlayerStats>()

    let orphanedMatches = 0
    for (const match of this.matchRepository.getAll()) {
      const gameType = this.gameTypeRepository.findById(match.gameTypeId)
      if (!gameType) {
        orphanedMatches++
        continue
      }
      const winners = new Set(getWinners(match, gameType))

      for (const ps of match.playerScores) {
        const current = stats.get(ps.playerId) ?? { playerId: ps.playerId, wins: 0, losses: 0 }
        stats.set(
          ps.playerId,
          winners.has(ps.playerId)
            ? { ...current, wins: current.wins + 1 }
            : { ...current, losses: current.losses + 1 },
        )
      }
    }
    if (orphanedMatches > 0) {
      console.warn(`[Scoreo] Warning: ${orphanedMatches} match(es) reference non-existent game types`)
    }
    return stats
  }
}
